### PR TITLE
Use temp filename while writing ImageLayer file

### PR DESCRIPTION
**Pull request by lubennikovaav**
#1740

I could not reproduce the #1735 test failure on my laptop.
@hlinnaka , could you test if this change fixes it?

### CODE REVIEW

**Review by hlinnaka — APPROVED** [review]
LGTM as far as this goes. We should add the fsync, but that's another pre-existing bug, not this PR's fault.

File: pageserver/src/layered_repository/image_layer.rs [hunk 1]
```diff
@@ -512,6 +524,22 @@ impl ImageLayerWriter {
...
+        let final_path = ImageLayer::path_for(
+            &PathOrConf::Conf(self.conf),
+            self.timelineid,
+            self.tenantid,
+            &ImageFileName {
+                key_range: self.key_range.clone(),
+                lsn: self.lsn,
+            },
+        );
+        std::fs::rename(self.path, &final_path)?;
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/layered_repository/image_layer.rs:541]: Oh I see. To recap, delta files are currently fsync'd. But only after renaming, and that's a problem.

File: pageserver/src/layered_repository/delta_layer.rs [hunk 2]
```diff
@@ -607,12 +622,9 @@ impl DeltaLayerWriter {
...
         // Note: This overwrites any existing file. There shouldn't be any.
         // FIXME: throw an error instead?
-        let path = conf.timeline_path(&timelineid, &tenantid).join(format!(
-            "{}-XXX__{:016X}-{:016X}.temp",
-            key_start,
-            u64::from(lsn_range.start),
-            u64::from(lsn_range.end)
-        ));
+        let path =
+            DeltaLayer::temp_path_for(conf, timelineid, tenantid, key_start, lsn_range.clone());
```
Comment (aome510) [inline pageserver/src/layered_repository/delta_layer.rs:626]: We can just pass `lsn_range` by reference and remove the `clone` here.

File: pageserver/src/layered_repository/delta_layer.rs [hunk 3]
```diff
@@ -420,6 +420,21 @@ impl DeltaLayer {
...
 
+    fn temp_path_for(
+        conf: &PageServerConf,
+        timelineid: ZTimelineId,
+        tenantid: ZTenantId,
+        key_start: Key,
+        lsn_range: Range<Lsn>,
+    ) -> PathBuf {
+        conf.timeline_path(&timelineid, &tenantid).join(format!(
+            "{}-XXX__{:016X}-{:016X}.temp",
```
Comment (aome510) [inline pageserver/src/layered_repository/delta_layer.rs:431]: We can also resolve https://github.com/neondatabase/neon/issues/1739 in this PR by appending a random string to the temp file name.

This, IMHO, is similar to how many temp file generators work.

File: pageserver/src/layered_repository/delta_layer.rs (hunk 2 above)
Comment (lubennikovaav) [reply pageserver/src/layered_repository/delta_layer.rs]: thank you, fixed

File: pageserver/src/layered_repository/delta_layer.rs (hunk 3 above)
Comment (lubennikovaav) [reply pageserver/src/layered_repository/delta_layer.rs]: Thank you for pointing out.
Fixed.

File: pageserver/src/layered_repository/delta_layer.rs [hunk 4]
```diff
@@ -420,6 +421,28 @@ impl DeltaLayer {
...
+        lsn_range: &Range<Lsn>,
+    ) -> PathBuf {
+        let rand_string: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(8)
+            .map(char::from)
+            .collect();
+
+        conf.timeline_path(&timelineid, &tenantid).join(format!(
+            "{}-XXX__{:016X}-{:016X}.temp.{}",
```
Comment (aome510) [inline pageserver/src/layered_repository/delta_layer.rs:438]:
```suggestion
            "{}-XXX__{:016X}-{:016X}.{}.temp",
```

It's better to keep the `temp` suffix, so we know which are temp files.

File: pageserver/src/layered_repository/image_layer.rs [hunk 5]
```diff
@@ -241,6 +242,22 @@ impl ImageLayer {
...
+        fname: &ImageFileName,
+    ) -> PathBuf {
+        let rand_string: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(8)
+            .map(char::from)
+            .collect();
+
+        conf.timeline_path(&timelineid, &tenantid)
+            .join(format!("{}.temp.{}", fname, rand_string))
```
Comment (aome510) [inline pageserver/src/layered_repository/image_layer.rs:258]: Same here

Comment (lubennikovaav) [reply pageserver/src/layered_repository/image_layer.rs]: Fair enough.
While on it, I also implemented the file cleanup to close #1732

**Review by aome510 — APPROVED** [review]
LGTM! Thanks for all the updates and fixes. Let's wait @hlinnaka for another approval before merge.

**Review by hlinnaka — APPROVED** [review]
So this now fixes github issues: #1740, #1739, and #1732.

LGTM, thanks!

**Pull request merged**